### PR TITLE
Replace maths functions with branches with assembly

### PR DESCRIPTION
**Pull request by TomAFrench**
These functions are far enough away from the business logic that I'm comfortable with making them less readable like this.

Most of these optimisations are to do with removing branches from the code to avoid JUMPs. As we're breaking type safety in assembly we can treat boolean values as a uint256 and then do maths with it directly.

One of the functions (`FixedPoint.mulUp`) goes _up_ in cost based on the snapshot. This is due to the new version being worse when `a == 0`, which I'd say is worth it for being cheaper the rest of the time.

### CODE REVIEW

**Comment by EndymionJkb** [comment]
This is great savings, but I think we should retain the original code in comments above it. That way we're not sacrificing readability either. (And people can independently verify the correctness of the assembly.)

File: pkg/solidity-utils/test/foundry/FixedPoint.t.sol [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            FixedPoint.divDown(a, b);
+        } else if (a != 0 && (a * FixedPoint.ONE) / FixedPoint.ONE != a) {
+            vm.expectRevert("BAL#005"); // DIV_INTERNAL
+            FixedPoint.divDown(a, b);
+        } else {
+            uint256 referenceDivDown = a == 0 ? 0 : (a * FixedPoint.ONE) / b;
+            uint256 divDown = FixedPoint.divDown(a, b);
+
+            assertEq(divDown, referenceDivDown);
+        }
```
Comment (nventuro) [inline pkg/solidity-utils/test/foundry/FixedPoint.t.sol:41]:
```suggestion
        if (b == 0) {
            vm.expectRevert("BAL#004"); // ZERO_DIVISION
        } else if (a != 0 && (a * FixedPoint.ONE) / FixedPoint.ONE != a) {
            vm.expectRevert("BAL#005"); // DIV_INTERNAL
        }

        uint256 divDown = FixedPoint.divDown(a, b);
        uint256 referenceDivDown = a == 0 ? 0 : (a * FixedPoint.ONE) / b;

        assertEq(divDown, referenceDivDown);
```

File: pkg/solidity-utils/test/foundry/Math.t.sol [hunk 2]
```diff
@@ -11,19 +11,30 @@ contract MathTest is Test {
...
     function testMax(uint256 a, uint256 b) external {
-        uint256 referenceMax = (a < b) ? b : a;
         uint256 max = Math.max(a, b);
+        assertGe(max, a);
+        assertGe(max, b);
+        assertTrue((max == a) || (max == b));
 
+        uint256 referenceMax = (a < b) ? b : a;
         assertEq(max, referenceMax);
     }
```
Comment (nventuro) [inline pkg/solidity-utils/test/foundry/Math.t.sol:29]: I would've split this into two tests - one comparing against the ref, and one testing props

File: pkg/solidity-utils/test/foundry/FixedPoint.t.sol (hunk 1 above)
Comment (TomAFrench) [reply pkg/solidity-utils/test/foundry/FixedPoint.t.sol:41]: I don't think I agree with this. IMO, `vm.expectRevert` should be kept as close as possible to the call we expect to blow up.

Comment (nventuro) [reply pkg/solidity-utils/test/foundry/FixedPoint.t.sol:41]: Fair enough

**Review by nventuro — APPROVED** [review]
:scared:

**Pull request merged**